hinted handoff: introduce HTTP API for waiting for hint replay (stateless version)

piodul opened the pull request:
## What it does

This PR introduces a new feature to hinted handoff: ability to wait until hints from given node are replayed towards a chosen set of nodes.

It replaces the old mechanism which waits for hints to be replayed before repair and exposes it through an HTTP API. The implementation is completely different, so this PR begins with a revert of the old functionality and then introduces the new implementation.

## The API

Waiting for hints is made possible with the help of "hint sync points". A sync point is a collection of positions in some hint queues from one node - those positions are encoded into the sync point's description as a hexadecimal string. The sync point consists only of its hexadecimal description - there is no state kept on any of the nodes.

Two operations are available through the HTTP API:

- `/hints_manager/waiting_point` (POST) - _Create a sync point_. Given a set of `target_hosts`, creates a sync point which encodes positions currently at the end of all queues pointing to any of the `target_hosts`.
- `/hints_manager/waiting_point` (GET) - _Wait or check the sync point_. Given a description of a sync point, checks if the sync point was already reached. If you provide a non-zero `timeout` parameter and the sync point is not reached yet, this endpoint will wait until it the point reached or the timeout expires.

## How it is implemented

Hinted handoff uses the commitlog framework in order to store and replay hints. Each entry (here, a serialized hint) can be identified by a "replay position", which contains the ID of the segment containing the hint, and its position in the file. Replay positions are ordered with respect to segment ID and then position in the file; because segment IDs are assigned sequentially and entries are also written sequentially, this order corresponds to the chronological order in which hints were written. This order also corresponds to the order in which hints are replayed, provided that hint segments are processed starting with the one with the smallest ID first.

The main idea is to track the positions of both the most recently written hint, and the most recently replayed hint. When creating a hint sync point, the position of the last written hint is encoded; when the sync point is waited on, the hints manager waits until the last replayed position reached the position encoded in the sync point. The description of the sync point encodes positions on a per-hint queue basis - separately for each shard, destination endpoint and hint type (regular or MV).

Note: although hints manager destroys and re-creates commitlog instances, the ordering described above still works - the ID of the first segment assigned by the commitlog instance corresponds to the number of milliseconds since the epoch, so commitlog instances created by newer instances will have larger IDs.

### Interaction with rebalancing and resharding

Before the hints manager is enabled, it performs segment _rebalancing_: for a given endpoint, it makes sure that each shard gets roughly the same number of hint segments. For example, if there are 3 shards and shard 1 has 7 segments, then shard 0 will get 2 segments, shard 1: 3 segments, and shard 2: 2 segments. Apart from distributing the work evenly between shards on startup, it also handles the case when the node is resharded - if the number of shards is reduced, segments from removed shards will be redistributed to lower shards.

Because of the possibility of segments being moved between shards on restart, this makes accurate tracking of hint replay harder. In order to simplify the problem, this PR changes the order in which hint segments are replayed - segments from other shards (here called "foreign" segments) are replayed first, before any "local" segment from this shard. Foreign segments are treated as if they were placed before the 0 replay position - when waiting for a hint sync point, we will __always__ wait for foreign segments to be replayed.

This behavior makes sure that hints generated before the sync point was created will be replayed - and, if segment rebalancing happened in the meantime, we will potentially replay some more segments which were moved across shards.

## Additional implementation remarks

This PR starts with a revert of the "hints: delay repair until hints are replayed" (#8452) functionality. Some infrastructure introduced in the original PR started to be used by other places in the code, so this is not a simple revert of the merge commit - instead, commits of the old PR are reverted separately and modified in order to make the code compile.

The following commits from the original PR were omitted from the revert because the code introduced by them became used by other logic in repair:

- 0db45d1df555d4b749828a8288571e36499ab18b (repair: introduce abort_source for repair abort)
- 3a2d09b644b5c203cfeeac9dd0b05bd5960d7474 (repair: introduce abort_source for shutdown)
- 49f4a2f9689860b728c10284eb7ccc79743a2068 (repair: plug in waiting for hints to be sent before repair)

Refs: #8102
Fixes: #8727

Tests: unit(dev)

piodul commented:
This is a newer and hopefully better version of this PR: https://github.com/scylladb/scylla/pull/8728/

psarna commented on db/hints/sync_point.cc:
```diff
@@ -0,0 +1,144 @@
+}
+
+sync_point sync_point::decode(sstring_view s) {
+    // Decode from hex
```
I think that the invocation below is quite self-descriptive already and this comment can be removed :)

piodul replied:
Sure. Removed.

psarna commented on db/hints/sync_point.cc:
```diff
@@ -0,0 +1,144 @@
+namespace db {
+namespace hints {
+
+// TODO: Encode host ID so that two sync points won't get mixed up
```
It looks like host id is already included in the sync point (although not in per_manager_sync_point), is this comment up to date?

Also, just curious - do you include the host id just to avoid a situation when a client gets the sync point from one node, and then mistakenly tries to wait on it on another node?

piodul replied:
> It looks like host id is already included in the sync point (although not in per_manager_sync_point), is this comment up to date?

No, it's not up to date. I forgot to remove it after I implemented the host ID check. I'll remove the comment.

> Also, just curious - do you include the host id just to avoid a situation when a client gets the sync point from one node, and then mistakenly tries to wait on it on another node?

Yes. It doesn't make sense to use a sync point created on one node to wait on another node. This check is verified in the dtest.

psarna commented on db/hints/sync_point.hh:
```diff
@@ -0,0 +1,56 @@
+    /// \brief Decodes a regular hints and MV hints sync points from an encoded, textual form.
+    static sync_point decode(sstring_view s);
+
+    /// Description TODO
```
That particular TODO should be fixed right away :)

piodul replied:
Ugh, I thought I fixed all of the "Description TODO"s... Fixed.

psarna commented on db/hints/sync_point.cc:
```diff
@@ -0,0 +1,144 @@
+    // Decode from hex
+    bytes raw = from_hex(s);
+    if (raw.empty()) {
+        throw std::runtime_error("Invalid hint sync point");
```
Let's mention that it's invalid because it's empty to make future debugging easier

piodul replied:
Improved the error message

psarna commented on db/hints/sync_point.cc:
```diff
@@ -0,0 +1,144 @@
+    std::vector<sync_point::shard_rps> ret;
+
+    if (size_t(shard_count) * v1.addresses.size() != v1.flattened_rps.size()) {
+        throw std::runtime_error("Invalid hint sync point");
```
Let's mention some context - that the sizes don't match, it will make future debugging easier

piodul replied:
Improved the error message

psarna commented on db/hints/sync_point.cc:
```diff
@@ -0,0 +1,144 @@
+}
+
+static per_manager_sync_point_v1 encode_one_type_v1(unsigned shards, const std::vector<sync_point::shard_rps>& rps) {
+    per_manager_sync_point_v1 ret;
```
nitpick: looks like the expected size of `ret.flattened_rps` is known (or at least its lower bound is) - which means it makes sense to reserve the space to avoid multiple reallocations inside the vector.

piodul replied:
Done, now we are preallocating.

psarna commented on idl/hinted_handoff.idl.hh:
```diff
@@ -39,6 +39,18 @@ struct sync_point_check_response {
+    utils::UUID host_id;
+    uint16_t shard_count;
+    db::hints::per_manager_sync_point_v1 regular_sp;
+    db::hints::per_manager_sync_point_v1 mv_sp;
```
I would appreciate a comment explaining these names - right now I'm fresh enough with this code to mentally expand it to "materialized views sync point", but I might have a hard time doing so a couple of months later

piodul replied:
Done, added comments which, I hope, explain the meaning.

psarna commented on service/storage_proxy.cc:
```diff
@@ -5317,112 +5303,98 @@ const db::hints::host_filter& storage_proxy::get_hints_host_filter() const {
+    spoint.regular_per_shard_rps.resize(smp::count);
+    spoint.mv_per_shard_rps.resize(smp::count);
+    spoint.host_id = co_await db::system_keyspace::get_local_host_id();
+    co_await parallel_for_each(boost::irange<unsigned>(0, smp::count), [this, &target_hosts, &spoint] (unsigned shard) {
```
This `parallel_for_each` kind of looks like a manual implementation of `invoke_on_all` - maybe it will be cleaner to use this primitive instead?

piodul replied:
I couldn't use `invoke_on_all` here because that method is not const, while `this` in current context is const.

avikivity replied:
I suppose we need a const version of invoke_on_all(), but it will be a pain to add all those overloads.

piodul commented:
v2:
- Addressed @psarna 's comments
- Fixed one of the revert commits: "messaging_service: add verbs for hint sync points" - it did not remove the idl.hh file; it also didn't remove the `register_`/`unregister_`/`send_` functions for the verbs, their removal was accidentally included in a later commit

piodul commented:
@avikivity @vladzcloudius @asias @psarna review ping

vladzcloudius commented:
> @avikivity @vladzcloudius @asias @psarna review ping

Since v2 addresses @psarna's comments we should probably wait for him to approve.

piodul commented:
@avikivity @asias ping

piodul commented:
@avikivity @asias ping^2

avikivity commented on api/api-doc/hinted_handoff.json:
```diff
@@ -7,6 +7,61 @@
+         "operations":[
+            {
+               "method":"POST",
+               "summary":"Creates a hints sync point. It can be used to wait until hints between given nodes are replayed. A sync point allows you to wait for hints accumulated at the moment of its creation - it won't wait for hints generated later.",
```
Please add an explanation here about how the sync point gets garbage collected later.

(real reason is that I'm too lazy to go back and hunt for it myself)

piodul replied:
Done (mentioned that sync point is described entirely by its ID and no state is kept on the Scylla side)

avikivity replied:
Wow, that is nice. Maybe I knew it before but I managed to forget it.

piodul replied:
Some context: https://github.com/scylladb/scylla/pull/8728#issuecomment-865026723

avikivity replied:
Ah, read that comment but then missed everything that followed, so it did not stick in my mind.

avikivity commented on db/hints/sync_point.cc:
```diff
@@ -0,0 +1,166 @@
+namespace db {
+namespace hints {
+
+// Format V1 (encoded as a hexadecimal string):
```
base64 will result in smaller things, we have functions for it.

piodul replied:
We do have functions for encoding in base64, but I was hesitant to use it because they are defined in the alternator directory. It feels strange for hints to depend on parts of the alternator, but maybe it's not harmful. I can switch to base64.

psarna replied:
@piodul feel free to move them out. It already happened with our rapidjson wrappers.

nyh replied:
Looks good, but we just need to remember in the future not to modify these functions in a way that can break Alternator. We do have a few tests in test/alternator which use "bytes" types (that use this base64 encoding) that should hopefully break if these utilities break - but I'm not sure how much of base64 corner cases these tests actually check (it certainly checks some, but possibly not all). We should have written a unit test in C++ directly for the base64 encoding and decoding functions that attempts to have full code coverage - or even better, use @denesb 's code coverage analyzer to make sure that test/alternator cover 100% of the base64 code.

Another minor problem with the base64.{hh,cc} is that its API is was designed to fit Alternator's needs exactly - in particular base64_encode returns std::string instead of seastar::sstring because this is what Altenator uses (for the reason why I want to move away from sstring, see https://github.com/scylladb/seastar/issues/634). It also has some peculiar APIs involving the subtly different "bytes" and "strings".

avikivity commented on test/manual/hint_test.cc:
```diff
@@ -44,6 +44,7 @@
 #include "db/commitlog/rp_set.hh"
 #include "log.hh"
 #include "schema.hh"
+#include "db/hints/sync_point.hh"
```
Why test/manual? Can it not run as part of the automated suite?

piodul replied:
You are right, it should be automatic. I put the new test in this file because it was the only file with hinted handoff unit tests. I'm not sure why the existing test is not run automatically, I'll do some git-archaeology and try to find out why it is the case.

piodul replied:
@vladzcloudius it looks like the `hint_test.cc` was never run automatically - although it was added to `configure.py` so that it is compiled (see [here](https://github.com/scylladb/scylla/commit/b20dbe16d84a3c1d1c77de90ee3cff6bbf2dc044)), it wasn't added to the list of test that should be run automatically in `test.py` (see [here](https://github.com/scylladb/scylla/blob/b20dbe16d84a3c1d1c77de90ee3cff6bbf2dc044/test.py#L30-L101)). Do you remember if it was intentional? It doesn't seem to be, so I'll preemptively move it to automatic tests in this PR - but I can move it back if you object.

avikivity replied:
You can add a new test file in the automatic section, and later move the manual test. This way we don't add a new blocker.

piodul replied:
Done, added another `hint_test.cc` file in `test/boost/` and placed the new test there.

piodul commented:
v3:
- Moved the base64 code to `utils`, and now sync points use base64 for encoding
- Moved the sync point serialization test to a new file `test/boost/hint_test.cc`

piodul commented:
@avikivity ping

piodul commented:
v4: Addressed an `#include`-related merge conflict in commit [`Revert "repair: plug in waiting for hints to be sent before repair" `](https://github.com/scylladb/scylla/pull/8982/commits/ecf854affcab4ffe0cd7fca3dfb2472f7c635fdf)

piodul commented:
@avikivity ping^2 (sorry for delayed ping, I was on vacation)
